src/communication/v2/session.hpp

```diff
@@ -230,12 +368,17 @@ class Session final : public std::enable_shared_from_this<Session<TSession, TSes
       parser.put(boost::asio::buffer(input_buffer_.read_end()->data(), input_buffer_.read_end()->size()),
                  error_code_parsing);
       if (!error_code_parsing) {
+        execution_active_ = false;
         if (boost::beast::websocket::is_upgrade(parser.get())) {
           spdlog::info("Switching {} to websocket connection", remote_endpoint_);
-          auto s = std::get<TCPSocket>(std::move(socket_));
-          socket_.emplace<WebSocket>(std::move(s));
-          std::get<WebSocket>(socket_).accept();
-          return DoRead();
+          if (std::holds_alternative<TCPSocket>(socket_)) {
+            auto sock = std::get<TCPSocket>(std::move(socket_));
+            WebsocketSession<TSession, TSessionData>::Create(std::move(sock), data_, endpoint_, service_name_)
+                ->DoAccept(parser.release());
```

Comment:
Here a `WebsoketSession` is created. As the created shared_ptr is released after `DoAccept`, it will only stay alive because of the `shared_ptr`s saved in its event handlers. Nice!